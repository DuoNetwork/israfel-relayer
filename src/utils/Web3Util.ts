import {
	assetDataUtils,
	BigNumber,
	ContractWrappers,
	generatePseudoRandomSalt,
	orderHashUtils,
	RPCSubprovider,
	signatureUtils,
	SignedOrder,
	SignerType,
	Web3ProviderEngine
} from '0x.js';
import { schemas, SchemaValidator } from '@0xproject/json-schemas';
import { MnemonicWalletSubprovider } from '@0xproject/subproviders';
import { Web3Wrapper } from '@0xproject/web3-wrapper';
import * as CST from '../common/constants';
import { IRawOrder, IStringSignedOrder } from '../common/types';
import infura from '../keys/infura.json';
import util from './util';

export enum Wallet {
	None,
	Local,
	MetaMask,
	Ledger
}

export default class Web3Util {
	public contractWrappers: ContractWrappers;
	public web3Wrapper: Web3Wrapper;
	public wallet: Wallet = Wallet.None;
	public accountIndex: number = 0;

	constructor(window: any, live: boolean, mnemonic: string) {
		if (window && typeof window.web3 !== 'undefined') {
			this.web3Wrapper = new Web3Wrapper(window.web3.currentProvider);
			this.wallet = Wallet.MetaMask;
		} else {
			const pe = new Web3ProviderEngine();
			pe.addProvider(
				new RPCSubprovider(
					(live ? CST.PROVIDER_INFURA_MAIN : CST.PROVIDER_INFURA_KOVAN) +
						'/' +
						infura.token
				)
			);
			if (!window && mnemonic) {
				const mnemonicWallet = new MnemonicWalletSubprovider({
					mnemonic: mnemonic,
					baseDerivationPath: CST.BASE_DERIVATION_PATH
				});
				pe.addProvider(mnemonicWallet);
			}
			pe.start();
			this.web3Wrapper = new Web3Wrapper(pe);
			this.wallet = window ? Wallet.None : Wallet.Local;
		}

		this.contractWrappers = new ContractWrappers(this.web3Wrapper.getProvider(), {
			networkId: live ? CST.NETWORK_ID_MAIN : CST.NETWORK_ID_KOVAN
		});
	}

	public onWeb3AccountUpdate(onUpdate: (addr: string, network: number) => any) {
		if (this.wallet !== Wallet.MetaMask) return;

		const store = (this.web3Wrapper.getProvider() as any).publicConfigStore;
		if (store)
			store.on('update', () => {
				onUpdate(
					store.getState().selectedAddress || '',
					Number(store.getState().networkVersion || '')
				);
			});
	}

	public async getCurrentAddress(): Promise<string> {
		const accounts = await this.web3Wrapper.getAvailableAddressesAsync();
		return accounts[this.accountIndex] || CST.DUMMY_ADDR;
	}

	public getCurrentNetwork(): Promise<number> {
		return this.web3Wrapper.getNetworkIdAsync();
	}

	public async createRawOrder(
		userAddr: string,
		relayerAddr: string,
		makerAssetAddr: string,
		takerAssetAddr: string,
		makerAmt: number,
		takerAmt: number,
		expInSeconds: number
	): Promise<IRawOrder> {
		if (this.wallet !== Wallet.MetaMask) Promise.reject('cannot sign');
		const order = {
			senderAddress: relayerAddr.toLowerCase(),
			makerAddress: userAddr.toLowerCase(),
			takerAddress: relayerAddr.toLowerCase(),
			makerFee: new BigNumber(0),
			takerFee: new BigNumber(0),
			makerAssetAmount: Web3Wrapper.toBaseUnitAmount(new BigNumber(makerAmt), 18),
			takerAssetAmount: Web3Wrapper.toBaseUnitAmount(new BigNumber(takerAmt), 18),
			makerAssetData: assetDataUtils.encodeERC20AssetData(makerAssetAddr),
			takerAssetData: assetDataUtils.encodeERC20AssetData(takerAssetAddr),
			salt: generatePseudoRandomSalt(),
			exchangeAddress: this.contractWrappers.exchange.getContractAddress(),
			feeRecipientAddress: relayerAddr.toLowerCase(),
			expirationTimeSeconds: new BigNumber(expInSeconds)
		};

		const orderHash = orderHashUtils.getOrderHashHex(order);
		const signature = await signatureUtils.ecSignOrderHashAsync(
			this.web3Wrapper.getProvider(),
			orderHash,
			order.makerAddress,
			SignerType.Metamask
		);
		return {
			orderHash: orderHash,
			signedOrder: { ...order, signature }
		};
	}

	public static getRandomFutureDateInSeconds = () => {
		return new BigNumber(Date.now() + CST.TEN_MINUTES_MS).div(CST.ONE_SECOND_MS).ceil();
	};

	public static stringToBN = (value: string): BigNumber => {
		return new BigNumber(value);
	};

	public static fromWei = (value: BigNumber | string): number => {
		return Number(Web3Wrapper.toUnitAmount(new BigNumber(value), 18).valueOf());
	};

	public static toWei = (value: BigNumber | string): number => {
		return Number(Web3Wrapper.toWei(new BigNumber(value)).valueOf());
	};

	public static assetDataToTokenName = (assetData: string): string => {
		const tokenAddr = assetDataUtils.decodeERC20AssetData(assetData).tokenAddress;
		return CST.TOKEN_MAPPING[tokenAddr];
	};

	public static getSideFromSignedOrder = (
		order: SignedOrder | IStringSignedOrder,
		pair: string
	): string => {
		return Web3Util.assetDataToTokenName(order.takerAssetData) === pair.split('-')[0]
			? CST.DB_BID
			: CST.DB_ASK;
	};

	public static getPriceFromSignedOrder = (order: IStringSignedOrder, side: string): number => {
		const isBid = side === CST.DB_BID;
		return util.round(
			Web3Util.stringToBN(isBid ? order.makerAssetAmount : order.takerAssetAmount)
				.div(isBid ? order.takerAssetAmount : order.makerAssetAmount)
				.valueOf()
		);
	};

	public async validateOrder(signedOrder: SignedOrder): Promise<string> {
		const { orderSchema } = schemas;
		const { signature, ...order } = signedOrder;
		const validator = new SchemaValidator();
		if (!validator.validate(order, orderSchema).valid) {
			util.logDebug('invalid schema ' + JSON.stringify(signedOrder));
			return '';
		}

		const orderHash = orderHashUtils.getOrderHashHex(order);
		const isValidSig = await signatureUtils.isValidSignatureAsync(
			this.web3Wrapper.getProvider(),
			orderHash,
			signature,
			order.makerAddress
		);
		if (!isValidSig) {
			util.logDebug('invalid signature ' + orderHash);
			return '';
		}

		return orderHash;
	}

	public getTokenAddressFromName(tokenName: string): string {
		switch (tokenName) {
			case CST.TOKEN_ZRX:
				return this.contractWrappers.exchange.getZRXTokenAddress();
			case CST.TOKEN_WETH:
				const ethTokenAddr = this.contractWrappers.etherToken.getContractAddressIfExists();
				if (!ethTokenAddr) {
					util.logInfo('no eth token address');
					return '';
				} else return ethTokenAddr;

			default:
				util.logInfo('no such token found');
				return '';
		}
	}

	public setAllUnlimitedAllowance(tokenAddr: string, addrs: string[]): Array<Promise<string>> {
		return addrs.map(address =>
			this.contractWrappers.erc20Token.setUnlimitedProxyAllowanceAsync(tokenAddr, address)
		);
	}

	public async setBaseQuoteAllowance(
		baseTokenAddr: string,
		quoteTokenAddr: string,
		addrs: string[]
	): Promise<void> {
		const responses = await Promise.all(
			this.setAllUnlimitedAllowance(quoteTokenAddr, addrs).concat(
				this.setAllUnlimitedAllowance(baseTokenAddr, addrs)
			)
		);
		await Promise.all(responses.map(tx => this.web3Wrapper.awaitTransactionSuccessAsync(tx)));
	}
}
