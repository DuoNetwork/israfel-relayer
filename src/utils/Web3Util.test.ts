// fix for @ledgerhq/hw-transport-u2f 4.28.0
import '@babel/polyfill';

import { BigNumber } from 'bignumber.js';
import Web3Util from './Web3Util';

// jest.mock('web3-utils');
jest.mock('@0x/subproviders', () => ({
	MetamaskSubprovider: jest.fn(),
	MnemonicWalletSubprovider: jest.fn()
}));

jest.mock('0x.js', () => ({
	ContractWrappers: jest.fn(),
	Web3ProviderEngine: jest.fn(
		() =>
			({
				addProvider: jest.fn(),
				start: jest.fn()
			} as any)
	),
	RPCSubprovider: jest.fn(),
	BigNumber: BigNumber
}));

jest.mock('@0x/contract-addresses', () => ({
	getContractAddressesForNetworkOrThrow: jest.fn()
}));

jest.mock('@0x/web3-wrapper', () => ({
	Web3Wrapper: jest.fn(() => ({
		getProvider: jest.fn(() => 'provider'),
		toBaseUnitAmount: jest.fn(value => value * 1e18)
	})),
	toBaseUnitAmount: jest.fn(value => value * 1e18)
}));

jest.mock('@0x/contract-addresses', () => ({
	getContractAddressesForNetworkOrThrow: jest.fn(() => 'contractAddr')
}));

jest.mock('web3-eth');
jest.mock('web3-eth-accounts');
jest.mock('web3-eth-personal');

test('constructor, with window, metaMask', () => {
	const window = {
		web3: {
			currentProvider: 'provider'
		}
	} as any;
	const testWeb3Util = new Web3Util(window, false, 'mnemonic', false);
	expect(testWeb3Util.wallet).toMatchSnapshot();
});

test('constructor, no window, local', () => {
	const testWeb3Util = new Web3Util(null, false, 'mnemonic', false);
	expect(testWeb3Util.wallet).toMatchSnapshot();
});
