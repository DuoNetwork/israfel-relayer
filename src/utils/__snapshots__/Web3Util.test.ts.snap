// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`awaitTransactionSuccessAsync 1`] = `
Array [
  Array [
    "txHash",
    undefined,
    undefined,
  ],
]
`;

exports[`constructor, no window, local 1`] = `1`;

exports[`constructor, no window, non local 1`] = `1`;

exports[`constructor, no window, non local, live 1`] = `1`;

exports[`constructor, no window, non local, no mnemonic 1`] = `0`;

exports[`constructor, with window, metaMask 1`] = `2`;

exports[`constructor, with window, metaMask, live 1`] = `2`;

exports[`getAvailableAddresses 1`] = `
Array [
  "addr1",
  "addr2",
]
`;

exports[`getCurrentAddress 1`] = `"addr1"`;

exports[`getCurrentNetwork 1`] = `42`;

exports[`getEthBalance 1`] = `
Array [
  Array [
    "address",
  ],
]
`;

exports[`getFilledTakerAssetAmount 1`] = `
Array [
  Array [
    "orderHash",
    Object {
      "defaultBlock": "latest",
    },
  ],
]
`;

exports[`getGasPrice 1`] = `1000000000`;

exports[`getProvider 1`] = `"provider"`;

exports[`getTokenAddressFromCode 1`] = `"0xd0a1e359811322d97991e03f863a0c30c2cf029c"`;

exports[`getTokenAddressFromCode, other token 1`] = `"otherTokenAddress"`;

exports[`getTokenAllowance, no spender 1`] = `
Array [
  Array [
    "tokenAddress",
    "owneraddr",
  ],
]
`;

exports[`getTokenAllowance, no tokenAddress 1`] = `0`;

exports[`getTokenAllowance, with spender 1`] = `
Array [
  Array [
    "tokenAddress",
    "owneraddr",
    "spender",
  ],
]
`;

exports[`getTokenBalance, no tokenAddress 1`] = `0`;

exports[`getTokenBalance, with tokenAddress 1`] = `
Array [
  Array [
    "tokenAddress",
    "address",
  ],
]
`;

exports[`getTokenByCode 1`] = `
Object {
  "code": "code",
}
`;

exports[`getTransactionCount 1`] = `10`;

exports[`getTransactionReceipt 1`] = `
Array [
  Array [
    "txHash",
  ],
]
`;

exports[`matchOrders 1`] = `
Array [
  Array [
    Object {
      "leftOrder": "leftOrder",
    },
    Object {
      "rightOrder": "rightOrder",
    },
    "senderAddr",
    Object {},
  ],
]
`;

exports[`setTokens 1`] = `
Array [
  Object {
    "address": "address",
    "code": "code1",
    "custodian": "custodian",
    "denomination": 1,
    "feeSchedules": Object {
      "WETH": Object {
        "minimum": 0.1,
        "rate": 0,
      },
    },
    "precisions": Object {
      "WETH": 0.0001,
    },
  },
  Object {
    "address": "address",
    "code": "code2",
    "custodian": "custodian",
    "denomination": 1,
    "feeSchedules": Object {
      "WETH": Object {
        "minimum": 0.1,
        "rate": 0,
      },
    },
    "precisions": Object {
      "WETH": 0.0001,
    },
  },
]
`;

exports[`setUnlimitedTokenAllowance, no spender 1`] = `
Array [
  Array [
    "tokenAddress",
    "account",
  ],
]
`;

exports[`setUnlimitedTokenAllowance, no tokenAddress 1`] = `"no tokenAddress"`;

exports[`setUnlimitedTokenAllowance, with spender 1`] = `
Array [
  Array [
    "tokenAddress",
    "account",
    "spender",
  ],
]
`;

exports[`tokenTransfer, sneder != from 1`] = `
Array [
  Array [
    "0xtokenAddr",
    "fromAddr",
    "toAddr",
    "senderAddr",
    undefined,
  ],
]
`;

exports[`tokenTransfer, sneder = from 1`] = `
Array [
  Array [
    "0xtokenAddr",
    "fromAddr",
    "toAddr",
    undefined,
  ],
]
`;

exports[`unwrapEther 1`] = `
Array [
  Array [
    "etherTokenAddress",
    10000000000000000000,
    "address",
  ],
]
`;

exports[`validateOrder 1`] = `"orderHash"`;

exports[`web3AccountsRecover 1`] = `
Array [
  Array [
    "message",
    "signature",
  ],
]
`;

exports[`web3AccountsRecover, no web3Accounts 1`] = `""`;

exports[`web3PersonalSign 1`] = `
Array [
  Array [
    "message",
    "account",
  ],
]
`;

exports[`web3PersonalSign, reject 1`] = `"canot sign"`;

exports[`wrapEther 1`] = `
Array [
  Array [
    "etherTokenAddress",
    10000000000000000000,
    "address",
  ],
]
`;
